File: source c++/Organisms/Organism.cpp
#include "Organism.h"

int Organism::instancesCount_ = 0;

Organism::Organism(int strength, int initiative, int x, int y, World& world)
	:	strength_(strength),
		initiative_(initiative),
		position_(x, y),
		world_(world),
		isAlive_(true)
{
	instanceNumber_ = ++instancesCount_;
}

Organism::Organism(Loader & loader, World & world)
	:	world_(world)

{
	position_.x = loader.NextInt();
	position_.y = loader.NextInt();
	isAlive_ = loader.NextInt();
	instanceNumber_ = loader.NextInt();
	strength_ = loader.NextInt();
	initiative_ = loader.NextInt();

	instancesCount_ = std::max(instancesCount_, instanceNumber_);
}


Organism::~Organism()
{
}

void Organism::Save(Saver & saver)
{
	saver.Push(position_.x);
	saver.Push(position_.y);
	saver.Push(isAlive_);
	saver.Push(instanceNumber_);
	saver.Push(strength_);
	saver.Push(initiative_);

}

int Organism::GetStrength() const
{
	return strength_;
}

int Organism::GetInitiative() const
{
	return initiative_;
}

int Organism::GetInstanceNumber() const
{
	return instanceNumber_;
}

int Organism::GetX() const
{
	return position_.x;
}

int Organism::GetY() const
{
	return position_.y;
}

bool Organism::isAlive() const
{
	return isAlive_;
}

utils::Point Organism::GetPosition() const
{
	return position_;
}

void Organism::SetStrength(int strength)
{
	strength_ = strength;
}

void Organism::SetInitiative(int initiative)
{
	initiative_ = initiative;
}

void Organism::Draw(Console & console)
{
	console.Print(GetLetter());
}

World& Organism::GetWorld()
{
	return world_;
}

void Organism::SetPosition(utils::Point position)
{
	position_ = position;
}

void Organism::Die()
{
	isAlive_ = false;
	world_.AddToRemovalSet(position_);
}

File: source java/organism/Organism.java
package organism;

import java.awt.Color;
import java.util.Comparator;
import organism.manager.Organisms;
import utilities.position.Position;
import utilities.savingsystem.Loader;
import utilities.savingsystem.Saver;
import virtualworld.World;

public abstract class Organism
{
    private static int instancesCount = 0;
    
    private boolean isAlive_;
    private final int instanceNumber_;
    private int strength_;
    private int initiative_;
    private Position position_;
    private final World world_;
    
    public Organism(int strength, int initiative, Position position, World world)
    {   
        this.strength_ = strength;
        this.initiative_ = initiative;
        this.position_ = position;
        this.world_ = world;
        
        isAlive_ = true;
        instanceNumber_ = ++instancesCount;
    }
    
    public Organism(Loader loader, World world, Position position)
    {
        position_ = position;
        isAlive_ = loader.NextBoolean();
        instanceNumber_ = loader.NextInt();
        strength_ = loader.NextInt();
        initiative_ = loader.NextInt();
        world_ = world;
        
        if(instanceNumber_ > instancesCount)
            instancesCount = instanceNumber_;
    }
    
    public abstract void Act();
    public abstract CollisionResult Collide(Organism organism);
    public abstract Organisms GetEnum();
    
    public void Die()
    {
        isAlive_ = false;
        GetWorld().AddToRemovalSet(this);
    }
    
    public void SetStrength(int strength)
    {
        strength_ = strength;
    }
    
    public void SetInitiative(int initiative)
    {
        initiative_ = initiative;
    }
    
    public void SetPosition(Position newPosition)
    {
        position_ = newPosition;
    }
    
    public int GetStrength()
    {
        return strength_;
    }
    
    public int GetInitiative()
    {
        return initiative_;
    }
    
    public Position GetPosition()
    {
        return position_;
    }
    
    public int GetInstanceNumber()
    {
        return instanceNumber_;
    }
    
    public boolean isAlive()
    {
        return isAlive_;
    }
    
    protected World GetWorld()
    {
        return world_;
    } 
    
    public Color GetColor()
    {
        return Color.green;
    }

    public void Save(Saver saver) 
    {
       saver.Print(Boolean.toString(isAlive_));
       saver.Print(Integer.toString(instanceNumber_));
       saver.Print(Integer.toString(strength_));
       saver.Print(Integer.toString(initiative_));
    }
    
    public static Comparator<Organism> OrganismInitiativeComparator = new Comparator<Organism>(){

        @Override
        public int compare(Organism a, Organism b) 
        {
            if(a.GetInitiative() == b.GetInitiative())
                return Integer.compare(b.GetInstanceNumber(), a.GetInstanceNumber());
            return Integer.compare(b.GetInitiative(), a.GetInitiative());
        }  
    };
}
    
            
   
